content/articles/protobuf-instead-of-json-in-frontend.md

```diff
@@ -0,0 +1,107 @@
+The server will deliver the content in JSON or Protobuf based on the `Accept` header.
+
+### Server-side
+Let's have a quick look at the server, it might not be new if you have already used Protobuf before but it is helpful for the rest (like me, I only used Thrift before).  
+First we need to define our Protobuf schema, which is done in the [addressbook.proto](https://github.com/WeAreWizards/protojson/blob/master/addressbook.proto) file. Here is one of the model as example:
+
+```protobuf
+message Contact {
+  required string first_name = 1;
+  required string last_name = 2;
+  optional Address address = 3;
+  repeated Phone phone_numbers = 4;
+};
+```
+In this model, `first_name` and `last_name` are required (controversial, I'll come back to that later), the address model is optional and we can have 0, 1 or n `phone_numbers`.  
+Now, Python doesn't know about .proto files so we need to use the [python client](https://pypi.python.org/pypi/protobuf) to generate code that we can use. This is the [addressbook_pb2.py](https://github.com/WeAreWizards/protojson/blob/master/addressbook_pb2.py) file. You can then import that in your Python code and access the models defined in your .proto file.  
+Have a look on the [data.py](https://github.com/WeAreWizards/protojson/blob/master/data.py) client to see an example. You can use those models as you would with a class and using kwargs for every attributes.  
+You also need to serialize them before sending, using the `SerializeToString()` method in our app.  
+
+### Client-side
+The client is written in Angular purely as an example and because ng-repeat was exactly what I wanted here. The code is a quick prototype so don't expect it to be idiomatic.  
+It uses [ProtoBuf.js](https://github.com/dcodeIO/ProtoBuf.js) as a Protobuf client.  
+The first thing we do is a basic GET to get all our contacts which is trivial since you can just dump the response data into our scope variable. On the other hand, we need to handle Protobuf deserialization since looking at our network tab we receive binary data looking like:
+
+```
+.
+MarlonSheets
+7771 Eastern Avenue11239
+,
+HarleyMalloy60139"
+(712)614-9303
+Y
+Mack  Hendricks3
+9573 Kimridge Cove
+7490 Orchard Hill Cove11239"
+(802)466-7004
+
+MadelineBowen87023
+0
+WillaSadler
+693 Burlington Parkway43824
+T
+```
+By default, this is more compact than JSON by quite a bit but once I turned on GZIP, the difference became much smaller: 851B for ProtoBuf and 942B for JSON.  
+
+In the example I copied the .proto file in the static directory and created our [models](https://github.com/WeAreWizards/protojson/blob/master/static/main.js#L12-L15) following the ProtoBuf.js tutorial.  
+Now the only two things left to do for our GET is to add a `responseType: 'arraybuffer'` in our HTTP requests (which would be done in the config for a proper angular app) and decode the data we received using `MyModel.decode(data)`.  
```

Comment:
I think we're still in the present tense  : "and decode the data we receive using"